File: SpringEcomAI/src/main/java/com/telusko/service/ProductService.java
package com.telusko.service;

import com.telusko.model.Product;
import com.telusko.repo.ProductRepo;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.chat.model.Generation;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.ai.chat.prompt.PromptTemplate;
import org.springframework.ai.converter.BeanOutputConverter;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.SearchRequest;
import org.springframework.ai.vectorstore.VectorStore;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.core.io.ResourceLoader;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.util.*;

@Service
public class ProductService {

    @Autowired
    private ProductRepo productRepo;

    @Autowired
    private ChatClient chatClient;

    @Autowired
    private VectorStore vectorStore;

    @Autowired
    private ResourceLoader resourceLoader;

    @Autowired
    private AIImageGeneratorService aiImageGeneratorService;

    public List<Product> getAllProducts() {
        return productRepo.findAll();
    }

    public Product getProductById(int id) {
        return productRepo.findById(id).orElse(new Product(-1));
    }

    public Product addOrUpdateProduct(Product product, MultipartFile image) throws IOException {

        if (image != null && !image.isEmpty()) {
            product.setImageName(image.getOriginalFilename());
            product.setImageType(image.getContentType());
            product.setProductImage(image.getBytes());

        }

        Product savedProduct = productRepo.save(product);

        String contentToEmbed =String.format("""
         Product Name: %s
         Description: %s
         Brand: %s
         Category: %s
         Price: %.2f
         Release Date: %s
         Available: %s
         Stock: %d
        """,
                savedProduct.getName(),
                savedProduct.getDescription(),
                savedProduct.getBrand(),
                savedProduct.getCategory(),
                savedProduct.getPrice(),
                savedProduct.getReleaseDate(),
                savedProduct.isProductAvailable(),
                savedProduct.getStockQuantity()
        );

        Document document = new Document(
                UUID.randomUUID().toString(),
                contentToEmbed,
                Map.of("productId", String.valueOf(savedProduct.getId()))
        );

        vectorStore.add(List.of(document));

        return savedProduct;
    }

    public String generateDescription(String name,String category){
        String descPrompt = String.format("""
            Write a concise and professional product description for an e-commerce listing.

            Product Name: %s
            Category: %s

            Keep it simple, engaging, and highlight its primary features or benefits.
            Avoid technical jargon and keep it customer-friendly.
            Limit the description to 250 characters maximum.
            """, name, category);

        return Objects.requireNonNull(chatClient.prompt(descPrompt)
                        .call()
                        .chatResponse())
                .getResult()
                .getOutput()
                .getText();

    }

    public byte[] generateImage(String name,String category,String description) {
        String imagePrompt = String.format("""
                        Generate a highly realistic, professional-grade e-commerce product image.

                        Product Details:
                        - Category: %s
                        - Name: '%s'
                        - Description: %s

                        Requirements:
                          - Use a clean, minimalistic, white or very light grey background.
                          - Ensure the product is well-lit with soft, natural-looking lighting.
                          - Add realistic shadows and soft reflections to ground the product naturally.
                          - No humans, brand logos, watermarks, or text overlays should be visible.
                          - Showcase the product from its most flattering angle that highlights key features.
                          - Ensure the product occupies a prominent position in the frame, centered or slightly off-centered.
                          - Maintain a high resolution and sharpness, ensuring all textures, colors, and details are clear.
                          - Follow the typical visual style of top e-commerce websites like Amazon, Flipkart, or Shopify.
                          - Make the product appear life-like and professionally photographed in a studio setup.
                          - The final image should look immediately ready for use on an e-commerce website without further editing.
                        """, category, name, description);

        byte[] aiImage = aiImageGeneratorService.generateImage(imagePrompt);

        return aiImage;
    }

    public void deleteProduct(int id) {
        productRepo.deleteById(id);
    }

    public List<Product> semanticSearchProducts(String userQuery) {
        try {
            String promptTemplate = Files.readString(
                    resourceLoader.getResource("classpath:prompts/product-search-prompt.st")
                            .getFile()
                            .toPath()
            );

            String context = fetchSemanticContext(userQuery);

            Map<String, Object> variables = new HashMap<>();
            variables.put("userQuery", userQuery);
            variables.put("context", context);

            PromptTemplate template = PromptTemplate.builder()
                    .template(promptTemplate)
                    .variables(variables)
                    .build();

            Prompt prompt = new Prompt(template.createMessage());

            Generation generation = chatClient.prompt(prompt)
                    .call()
                    .chatResponse()
                    .getResult();

            BeanOutputConverter<List<Product>> outputConverter = new BeanOutputConverter<>(
                    new ParameterizedTypeReference<>() {}
            );
            List<Product> aiProducts = outputConverter.convert(generation.getOutput().getText());

            List<Integer> productIds = aiProducts.stream()
                    .map(Product::getId)
                    .filter(id -> id > 0)
                    .toList();

            return productRepo.findAllById(productIds);

        } catch (IOException e) {
            throw new RuntimeException("Failed to process query", e);
        }
    }

    private String fetchSemanticContext(String query) {
        List<Document> documents = vectorStore.similaritySearch(
                SearchRequest.builder()
                        .query(query)
                        .topK(5)
                        .build()
        );

        StringBuilder contextBuilder = new StringBuilder();
        for (Document doc : documents) {
            contextBuilder.append(doc.getFormattedContent()).append("\n");
        }

        return contextBuilder.toString();
    }

}

File: SpringEcomAI-LangChain/services/product_service.py
from decimal import Decimal
import base64
import uuid
import json
import re
from pathlib import Path

from sqlalchemy.orm import Session
from langchain_openai import ChatOpenAI
from langchain_core.documents import Document

from models import Product
from config.vector_store import vector_store
from services.ai_image_service import generate_image

llm = ChatOpenAI(model="gpt-4o")

SEARCH_PROMPT_PATH = Path(__file__).parent.parent / "prompts" / "product-search-prompt.st"


def _load_search_prompt(context: str, user_query: str) -> str:
    template = SEARCH_PROMPT_PATH.read_text(encoding="utf-8")
    return template.replace("{userQuery}", user_query).replace("{context}", context)


def _to_snake_case(name: str) -> str:
    return re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()


def _product_to_text(product: Product) -> str:
    return f"""
Product Name: {product.name}
Description: {product.description}
Brand: {product.brand}
Category: {product.category}
Price: {float(product.price):.2f}
Release Date: {product.release_date}
Available: {product.product_available}
Stock: {product.stock_quantity}
""".strip()


def _delete_from_vector_store(product_id: int):
    collection = vector_store._collection
    results = collection.get(where={"productId": str(product_id)})
    if results["ids"]:
        vector_store.delete(ids=results["ids"])


def _index_product(product: Product):
    doc = Document(
        page_content=_product_to_text(product),
        metadata={"productId": str(product.id)}
    )
    vector_store.add_documents([doc], ids=[str(uuid.uuid4())])


def _convert_types(data: dict) -> dict:
    """Convert string values from React to proper Python types — same as Spring Boot Jackson."""

    data.pop("product_image", None)
    data.pop("image_name", None)
    data.pop("image_type", None)

    if "release_date" in data and isinstance(data["release_date"], str):
        try:
            from datetime import datetime
            data["release_date"] = datetime.strptime(data["release_date"], "%Y-%m-%d").date()
        except (ValueError, TypeError):
            data["release_date"] = None

    if "price" in data and data["price"] is not None:
        data["price"] = Decimal(str(data["price"]))

    if "stock_quantity" in data and data["stock_quantity"] is not None:
        data["stock_quantity"] = int(data["stock_quantity"])

    return data


def get_all_products(db: Session):
    return db.query(Product).all()


def get_product_by_id(db: Session, product_id: int):
    return db.query(Product).filter(Product.id == product_id).first()


def add_or_update_product(
    db: Session,
    product_data: dict,
    image_bytes: bytes = None,
    image_name: str = None,
    image_type: str = None
) -> Product:
    product_data = {_to_snake_case(k): v for k, v in product_data.items()}

    product_data = _convert_types(product_data)

    product_id = product_data.get("id")

    if product_id:
        product = db.query(Product).filter(Product.id == product_id).first()
        if not product:
            product = Product()
    else:
        product = Product()

    for key, value in product_data.items():
        if hasattr(product, key) and key != "id":
            setattr(product, key, value)

    if image_bytes:
        product.image_name = image_name
        product.image_type = image_type
        product.product_image = image_bytes

    db.add(product)
    db.commit()
    db.refresh(product)

    _delete_from_vector_store(product.id)
    _index_product(product)

    return product


def delete_product(db: Session, product_id: int):
    product = db.query(Product).filter(Product.id == product_id).first()
    if product:
        db.delete(product)
        db.commit()


def generate_description(name: str, category: str) -> str:
    prompt = f"""
Write a concise and professional product description for an e-commerce listing.

Product Name: {name}
Category: {category}

Keep it simple, engaging, and highlight its primary features or benefits.
Avoid technical jargon and keep it customer-friendly.
Limit the description to 250 characters maximum.
""".strip()
    return llm.invoke(prompt).content


def generate_product_image(name: str, category: str, description: str) -> bytes:
    prompt = f"""
Generate a highly realistic, professional-grade e-commerce product image.

Product Details:
- Category: {category}
- Name: '{name}'
- Description: {description}

Requirements:
  - Clean white/light grey background, soft natural lighting.
  - No humans, logos, watermarks, or text.
  - Professional e-commerce style like Amazon or Flipkart.
""".strip()
    return generate_image(prompt)


def semantic_search_products(db: Session, query: str):

    all_products = db.query(Product).all()
    
    context = "\n\n".join([_product_to_text(p) + f"\nProduct ID: {p.id}" for p in all_products])

    prompt = _load_search_prompt(context, query)

    response = llm.invoke(prompt).content
    response = re.sub(r"```json|```", "", response).strip()

    products_data = json.loads(response)
    product_ids = [int(p["id"]) for p in products_data if int(p.get("id", 0)) > 0]

    return db.query(Product).filter(Product.id.in_(product_ids)).all()